src/Bicep.Core/Emit/ExpressionEmitter.cs

```diff
@@ -162,10 +164,92 @@ public void EmitLanguageExpression(SyntaxBase syntax)
             writer.WriteValue(serialized);
         }
 
+        public void EmitCopyObject(string name, ForSyntax syntax, SyntaxBase? input, string? copyIndexOverride = null)
+        {
+            writer.WriteStartObject();
+
+            this.EmitProperty("name", name);
+            // construct the length ARM expression from the Bicep array expression
+            // type check has already ensured that the array expression is an array
+            this.EmitPropertyWithTransform(
+                "count",
+                syntax.Expression,
+                arrayExpression => new FunctionExpression("length", new[] { arrayExpression }, Array.Empty<LanguageExpression>()));
+
+            if (input != null)
+            {
+                if (copyIndexOverride == null)
+                {
+                    this.EmitProperty("input", input);
+                }
+                else
+                {
+                    this.EmitPropertyWithTransform("input", input, expression =>
+                    {
+                        // this is a giant hack
+                        // the named copy index in the serialized expression is incorrect
+                        // because the object syntax here does not match the JSON equivalent due to the presence of { "value": ... } wrappers
+                        // for now, we will manually replace the copy index in the converted expression
+                        // this also will not work for nested property loops
```

Comment:
What's the less-hacky solution? Will you create an issue for it?

Reply:
Will try it in this PR - basically use the local symbol replacement logic to do it instead of the `ExpressionVisitor`.

---
In reply to: [582201312](https://github.com/Azure/bicep/pull/1521#discussion_r582201312) [](ancestors = 582201312)

Reply:
Decided to leave as is for now. Will have to revisit when we support nested property loops.

---
In reply to: [582207976](https://github.com/Azure/bicep/pull/1521#discussion_r582207976) [](ancestors = 582207976,582201312)